### PR TITLE
mailman: support multiple oauth tokens.

### PR DESCRIPTION
**Pull request by gjoliver**

### CODE REVIEW

File: mailman/mailman.go [hunk 1]
```diff
@@ -20,44 +21,48 @@ import (
...
-	Token    Token
+	Tokens   Tokens
 }
 
 // New creates a new mailman.
 func New(c *Config) *Mailman {
 	states := signer.NewSessions(c.StateKey, time.Minute*3)
 
 	const gmailSendScope = "https://www.googleapis.com/auth/gmail.send"
+	const emailScope = "https://www.googleapis.com/auth/userinfo.email"
```
Comment (h8liu) [inline mailman/mailman.go:49]: let's just inline those now. The content of the strings are self explaining. Or just define `scopes := []string {...}` here and use `scopes` in the Config construction.

I defined the const mostly for just shortening the line (to limit it into 80 chars).



this is userInfoEmail scope..

next time, you can use a const block here:

`
const (
    gmailSendScope = ...
    emailScope = ...
)
`

File: mailman/mailman.go [hunk 2]
```diff
@@ -70,19 +75,28 @@ func (m *Mailman) tokenState(c *aries.C) (*oauth2.Token, *oauth.State, error) {
 }
 
 func (m *Mailman) serveIndex(c *aries.C) error {
-	tok, err := m.token.Get(c.Context)
+	emails, ok := c.Req.URL.Query()["email"]
+	if !ok {
+		c.Resp.Write([]byte("Please specify email parameter"))
```
Comment (h8liu) [inline mailman/mailman.go:80]: please return an error here:

`return errcode.InvalidArgf("email not specified in query")`

File: mailman/mailman.go [hunk 3]
```diff
@@ -70,19 +75,28 @@ func (m *Mailman) tokenState(c *aries.C) (*oauth2.Token, *oauth.State, error) {
 }
 
 func (m *Mailman) serveIndex(c *aries.C) error {
-	tok, err := m.token.Get(c.Context)
+	emails, ok := c.Req.URL.Query()["email"]
```
Comment (h8liu) [inline mailman/mailman.go:78]: use `Query().Get("email")`

File: mailman/mailman.go (hunk 2 above)
Comment (h8liu) [reply mailman/mailman.go:80]: the error message will be printed to Resp's body.

also, next time you want to print a string to a Writer, use `io.WriteString()`

File: mailman/mailman.go [hunk 4]
```diff
@@ -70,19 +75,28 @@ func (m *Mailman) tokenState(c *aries.C) (*oauth2.Token, *oauth.State, error) {
...
+		c.Resp.Write([]byte("Please specify email parameter"))
+		return nil
+	}
+
+	// We simply take the first specified email parameter.
+	tok, err := m.tokens.Get(c.Context, emails[0])
 	if err != nil {
 		if errcode.IsNotFound(err) {
-			return fmt.Errorf("mailman token not found")
+			c.Resp.Write([]byte("Token not found"))
```
Comment (h8liu) [inline mailman/mailman.go:88]: why this change here?

File: mailman/mailman.go [hunk 5]
```diff
@@ -70,19 +75,28 @@ func (m *Mailman) tokenState(c *aries.C) (*oauth2.Token, *oauth.State, error) {
...
 		return err
 	}
 	return aries.PrintJSON(c, tok)
 }
 
 // Send sends an email. Needs to setup OAuth2 first.
-func (m *Mailman) Send(ctx context.Context, body []byte) (string, error) {
-	tok, err := m.token.Get(ctx)
+func (m *Mailman) Send(
+	ctx context.Context, from string, body []byte) (string, error) {
```
Comment (h8liu) [inline mailman/mailman.go:98]: nit: break line like this:
```
func (m *Mailman) Send(
   ctx context.Context, from string, body []byte,
) (string, error) {
   ...
}
```

File: mailman/mailman.go [hunk 6]
```diff
@@ -137,11 +152,27 @@ func (m *Mailman) serveCallback(c *aries.C) error {
 	if token.RefreshToken == "" {
 		return fmt.Errorf("refresh token empty")
 	}
-	if err := m.token.Set(c.Context, token); err != nil {
+
+	// Let's find out which email this callback is for.
```
Comment (h8liu) [inline mailman/mailman.go:156]: nit: remove the "Let's" in the comment (like who is us here?), maybe say: "Get the user's email address."

File: mailman/mailman.go [hunk 7]
```diff
@@ -137,11 +152,27 @@ func (m *Mailman) serveCallback(c *aries.C) error {
...
+	const userInfoURL = "https://www.googleapis.com/oauth2/v3/userinfo"
```
Comment (h8liu) [inline mailman/mailman.go:157]: this is also used in the oauth package. could you generalize this API call into a function and use it for both usage so that it is not repeated?

File: mailman/mailman.go [hunk 8]
```diff
@@ -137,11 +152,27 @@ func (m *Mailman) serveCallback(c *aries.C) error {
...
+		return err
+	}
+
+	if err := m.tokens.Set(c.Context, user.Email, token); err != nil {
 		return err
 	}
 
-	c.Redirect(path.Dir(c.Path)) // redirect to index
+	// redirect to index with email parameter
+	c.Redirect(path.Dir(c.Path) + "?email=" + user.Email)
```
Comment (h8liu) [inline mailman/mailman.go:175]: please form the URL string with `url.URL.String()`, so that it is properly escaped.

like '@' needs to be escaped here. example: https://play.golang.org/p/NHRNBGZmZZq

**Review by h8liu — CHANGES_REQUESTED** [review]

File: mailman/mailman.go (hunk 1 above)
Comment (gjoliver) [reply mailman/mailman.go:49]: done

File: mailman/mailman.go (hunk 2 above)
Comment (gjoliver) [reply mailman/mailman.go:80]: ok

File: mailman/mailman.go (hunk 3 above)
Comment (gjoliver) [reply mailman/mailman.go:78]: done

File: mailman/mailman.go (hunk 4 above)
Comment (gjoliver) [reply mailman/mailman.go:88]: reverted

File: mailman/mailman.go (hunk 5 above)
Comment (gjoliver) [reply mailman/mailman.go:98]: yeah, now I remembered.

File: mailman/mailman.go (hunk 6 above)
Comment (gjoliver) [reply mailman/mailman.go:156]: done

File: mailman/mailman.go (hunk 7 above)
Comment (gjoliver) [reply mailman/mailman.go:157]: done.
I am a bit confused as to whether you want me to create an API GetUserInfo on Client type for all the oauth clients, or just a simple util func in google.go.
did it the latter way, but let me know.

File: mailman/mailman.go (hunk 8 above)
Comment (gjoliver) [reply mailman/mailman.go:175]: done. nice.

**Comment by gjoliver** [comment]
Updated. PTAL.

File: mailman/mailman.go [hunk 9]
```diff
@@ -19,45 +19,47 @@ import (
...
+	Tokens   Tokens
 }
 
 // New creates a new mailman.
 func New(c *Config) *Mailman {
 	states := signer.NewSessions(c.StateKey, time.Minute*3)
 
-	const gmailSendScope = "https://www.googleapis.com/auth/gmail.send"
-
+	var scopes = []string{
```
Comment (h8liu) [inline mailman/mailman.go:47]: please use `:=` rather than `var`

use `var` when there are no init values (will zero all values in those cases).

File: oauth/google.go [hunk 10]
```diff
@@ -28,6 +29,11 @@ const (
 
 type google struct{ c *Client }
 
+type GoogleUserInfo struct {
```
Comment (h8liu) [inline oauth/google.go:32]: You need to add doc comment for this struct, otherwise, it won't be passing `smake`

File: oauth/google.go [hunk 11]
```diff
@@ -53,25 +59,33 @@ func newGoogle(app *GoogleApp, s *signer.Sessions) *google {
 
 func (g *google) client() *Client { return g.c }
 
+func GetGoogleUserInfo(
+	c *Client, ctx context.Context, tok *oauth2.Token,
```
Comment (h8liu) [inline oauth/google.go:63]: context is always the first argument, no matter what.

File: oauth/google.go [hunk 12]
```diff
@@ -53,25 +59,33 @@ func newGoogle(app *GoogleApp, s *signer.Sessions) *google {
...
+) (*GoogleUserInfo, error) {
+	bs, err := c.Get(ctx, tok, "https://www.googleapis.com/oauth2/v3/userinfo")
+	if err != nil {
+		return nil, err
+	}
+
+	var user GoogleUserInfo
```
Comment (h8liu) [inline oauth/google.go:70]: nit: use `user := new(GoogleUserInfo)` here, and directly use the pointer afterwards.

File: oauth/google.go (hunk 10 above)
Comment (h8liu) [reply oauth/google.go:32]: also, could you 1. put this struct together along with `GetGoogleUserInfo()` function. 2. put these two items at the very top or bottom of this file, so that they do not split the google *class* into chunks?

**Review by h8liu — CHANGES_REQUESTED** [review]
just some small things this time.

File: mailman/mailman.go (hunk 7 above)
Comment (h8liu) [reply mailman/mailman.go]: just a simple util func is fine.

File: mailman/mailman.go (hunk 9 above)
Comment (gjoliver) [reply mailman/mailman.go:47]: done

File: oauth/google.go (hunk 10 above)
Comment (gjoliver) [reply oauth/google.go:32]: done

File: oauth/google.go (hunk 11 above)
Comment (gjoliver) [reply oauth/google.go:63]: done

File: oauth/google.go (hunk 12 above)
Comment (gjoliver) [reply oauth/google.go:70]: done

**Comment by gjoliver** [comment]
ok. done. thanks.

File: oauth/google.go (hunk 10 above)
Comment (h8liu) [reply oauth/google.go:33]: put these before the `type google struct`

File: oauth/google.go [hunk 13]
```diff
@@ -28,6 +29,28 @@ const (
 
 type google struct{ c *Client }
 
+// GoogleUserInfo stores a Google user's basic personal info.
+type GoogleUserInfo struct {
+	Email string `json:"email"`
+	Name  string `json:"name"`
+}
+
+func GetGoogleUserInfo(
```
Comment (h8liu) [inline oauth/google.go:38]: doc comment?

please make sure `smake` runs without issue.

**Review by h8liu — APPROVED** [review]

File: oauth/google.go (hunk 13 above)
Comment (gjoliver) [reply oauth/google.go:37]: done.
always forgot to go into all the dependency directories and run smake there.
what's your workflow look like? can you let smake check any folders that latest commit touches or something?

Comment (h8liu) [reply oauth/google.go:37]: I basically use `smake` to compile. no matter what, you need to compile once the code before submitting a commit?
